File: spiralTraverse.java
// Time Complexity : O(n * m)
// Space Complexity : O(1)
// Did this code successfully run on Leetcode : Yes
// we've taken 4 variables to traverse thru the matrix in top, bottom, left and right 
//directions and were keeping track of either traversinng in the boundry or the internal
// part of the spiral 

class Solution {
    public List<Integer> spiralOrder(int[][] matrix) {
        int m = matrix.length;
        int n = matrix[0].length;
        List<Integer> result = new ArrayList<>();
        int top = 0;
        int bottom = m-1;
        int left = 0;
        int right = n-1;
        while(left <= right && top <= bottom){
            //top row
            for(int j=left; j <= right; j++){
                result.add(matrix[top][j]);
            }
            top++;
            //right column
            for(int i = top; i <= bottom; i++){
                result.add(matrix[i][right]);
            }
            right--;
            //bottom row
            if(left <= right && top <= bottom)
            for(int j = right; j >= left; j--){
                result.add(matrix[bottom][j]);
            }
            bottom--;

            //left column
            if(left <= right && top <= bottom)
            for(int i = bottom; i >= top; i--){
                result.add(matrix[i][left]);
            }
            left++;
        }
        return result;
    }
}

File: spiral_traverse.py
# // Time Complexity : O(m*n)
# // Space Complexity : O(m*n)
# // Did this code successfully run on Leetcode : Yes but tricky in python
# // Three line explanation of solution in plain english

# // Your code here along with comments explaining your approach

class Solution:
    def spiralOrder(self, matrix: List[List[int]]) -> List[int]:
        if len(matrix)==0 or matrix==None:
            return []
        
        m=len(matrix)
        n=len(matrix[0])
        left,top=0,0
        right,bottom=n,m
        result=[]
        
        while top<bottom and left<right:
            
            #Top Wall
            for i in range(left,right):
                result.append(matrix[top][i])
            
            top+=1
            
            #Right Wall
            for i in range(top,bottom):
                result.append(matrix[i][right-1])
            
            right-=1
            
            if not(left<right and top<bottom):
                break
            #Bottom Wall
            for i in range(right-1,left-1,-1):
                result.append(matrix[bottom-1][i])
            
            bottom-=1
            
            #left wall
            for i in range(bottom-1,top-1,-1):
                result.append(matrix[i][left])
            left+=1
        return result